Latest fixes as were used to post to LKML

captain5050 opened the pull request:
LKML post: https://lore.kernel.org/linux-perf-users/20221001060636.2661983-1-irogers@google.com/

Remove a workaround for SPR fixed in mapfile.csv but add in a new ADL one.
Topdown metrics are out of 100%, so add a ScaleUnit.
If topdown events are used, ensure the metric has a slots event to avoid a broken group (tested on Tigerlake).
Prefer generated metrics over ones from github as the units will match.
Fix some typing, code hygiene issues.

captain5050 committed:
Update mapfile.csv bug fixes

Modify if to be a loop to ensure 7 columns.
Fix Alderlake with only Atoms as perf won't support the core
architecture being both atom and core.

captain5050 committed:
All topdown metrics have a ScaleUnit of 100%

Prefer a percentage rather than a value between 0 and 1.

captain5050 committed:
Add option to expand all metrics

This makes the behavior of expanding metrics within the expression
configurable.

captain5050 committed:
If topdown events are used, insert a slots event

Topdown events need to be in a group with a slots event. Add a slots
event by by adding it multiplied by zero.

edwarddavidbaker commented on download_and_gen.py:
```diff
@@ -229,8 +229,9 @@ def __init__(self, base_url: str, metrics_url: str):
             for l in mapfile:
-                if len(l) == 6:
-                    l.append("")
+                while len(l) < 7:
```
Ian, Sorry about hiccups in the mapfile. Thanks for working around them! I've added a local check for both column count as well as uncore experimental.

captain5050 replied:
Thanks! I'm unclear what to do about the atom case with Alderlake CPUID 6-BE, will that change in the future? Perhaps we need to do something special for that CPU in Linux as normally on a hybrid system the PMU for atom is called cpu_atom, but maybe on this model it is cpu_core or just cpu?
Wrt local checks, you may want to check what is checked here that each model maps to a single path (except for hybrid where we expect exactly 2) and that things like core/uncore/experimental uncore agree with the what is in the path. There was previously an issue for this and SPR.
Something I've also raised is that there are fewer CPUIDs here compared to:
https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/intel-family.h#L45
and the perf tool has also listed CPUIDs not in this file previously (A7 and 7D both icelake variants). It looks like palm cove, sunny cove, lightning mountain and quark x1000 are the only missing cases now - likely intentionally :-)

edwarddavidbaker replied:
Hi Ian, Thanks for the feedback.

- I'm unclear what to do about the atom case with Alderlake CPUID 6-BE, will that change in the future? Perhaps we need to do something special for that CPU in Linux as normally on a hybrid system the PMU for atom is called cpu_atom, but maybe on this model it is cpu_core or just cpu?
  - I believe the current 0xBE mapfile entry is a correct representation of Alder Lake N (https://github.com/torvalds/linux/blob/534b0abc627a4034ecaf177e780b1680e61b37f4/arch/x86/include/asm/intel-family.h#L112) . Sorry about the additional complexity.
- Wrt local checks, you may want to check what is checked here that each model maps to a single path (except for hybrid where we expect exactly 2)
   - Thanks. I'll add a check.
- Things like core/uncore/experimental uncore agree with the what is in the path.
   - Definitely. I have checks for uncore vs uncore experimental since I broke Snow Ridge recently. I'll try to expand these types of checks soon.
- There are fewer CPUIDs here compared to:
https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/intel-family.h#L45
   - This was a great suggestion. As you've noticed the mapfile is updated with only a few missing. The current missing IDs are intentionally not added.

captain5050 replied:
Great! Is there anything outstanding from this pull request? Would be nice to land it.

captain5050 committed:
Prefer existing metrics over github

Previously a github metric would be preferred over one generated by
extract-tma-metrics.py. As these metrics now refer to other metrics,
the same metrics must be used to ensure the units are the same.
Remove broken_extra_metrics as no missing events were found in the
latest testing.

captain5050 committed:
Fix numerous type and lint issues

Prefer re.fullmatch over re.match.  Add missing types and use pytype
to detect issues. In extract-tma-metrics.py, make groups
non-ambiguous/global.
No impact on generated perf json.

captain5050 committed:
Merge branch 'intel:master' into master

ak-intel commented:
Thanks!
